unicode_literals for bytes to unicode.

rtgdk opened the pull request:
Fixes #11 
`from __future__ import unicode_literals ` will change bytes to unicode if run in Python2. In Python3, it won't have any affect. 
Please review it, if it's correct then I'll extend it to other files.

pombredanne commented:
@rtgdk Thanks. But please do not pile unrelated commits in the same pull request. Can you submit a separate request for your unicode related commit?

pombredanne commented on spdx/writers/tagvalue.py:
```diff
@@ -12,7 +12,7 @@
 from __future__ import absolute_import
 from __future__ import print_function
-
+from __future__ import unicode_literals
```
Please keep an empty line after __future__ import

rtgdk commented:
@pombredanne  I have now imported unicode_literals wherever necessary.

rtgdk commented:
@pombredanne  Any review on this ? Will this work ? 
FWIW my basic testing on this worked. Since unicode_literals will only affect Python2 bytes and convert them to unicode. str of Python3 is already unicode. So we don't have to write like `u"string"`  anymore.

pombredanne commented:
> Any review on this ? Will this work ?

Did you check the results on the CI tests? Are they passing everywhere on Py2 and 3?

rtgdk commented:
@pombredanne  After looking through the tests failure and code, I realised I imported unicode_literals unnecessarily in some files which led to conversion of some bytes(str) to unicode where files are being opened in binary mode to read/write.( An error in python 2)
I have rectified my error. The tests also seem to pass.

rtgdk commented:
@pombredanne Did you have a look at the changes?

pombredanne commented:
Thanks for the update!

1. To use unicode throughout, you want to import unicode_literals everywhere. not somewhere.
2. you need to add new tests in any case.

rtgdk commented:
But if I import unicode_literals everywhere then some of the current test are bound to fail. Like this `test_rdf_rdf` in `test_conversion.py` . Since the file is being opened with `open` and not with `codec.open`. Also the file is being opened in `rb` without any encoding/decoding. So it needs data in bytes and importing unicode_literals in those files will raise encoding/decoding error.
But if the file is being opened with codec.open in "w" with `encoding="utf-8"` all is working perfectly since all the strings are implicitly converted to unicode before writing.

pombredanne commented:
Yes,  you need to fix the code, fix existing tests and write new tests. 
Adding the unicode_literals import is only the simple first step.

pombredanne commented on spdx/document.py:
```diff
@@ -51,7 +52,7 @@ def from_full_name(cls, full_name):
         the identifier is the same as the full_name.
         """
-        if full_name in config.LICENSE_MAP.keys():
+        if full_name.encode("utf-8") in config.LICENSE_MAP.keys():
```
Based on my overall review comment, this should never be needed here. e.g. if you use unicode throughout, do any needed conversion at the boundaries and not here.

pombredanne commented on spdx/document.py:
```diff
@@ -40,7 +41,7 @@ def from_identifier(cls, identifier):
         the full_name is retrived from it. Otherwise
```
Do not hesitate to fix typos as you go: this should be "retrieved" , not "retrived"

pombredanne commented on spdx/parsers/loggers.py:
```diff
@@ -12,7 +12,7 @@
 from __future__ import absolute_import
 from __future__ import print_function
-
+from __future__ import unicode_literals
```
Always keep two empty lines after imports

pombredanne replied:
?

pombredanne commented on spdx/parsers/tagvalue.py:
```diff
@@ -507,7 +508,7 @@ def p_conc_license_3(self, p):
             value = p[1]
         ref_re = re.compile('LicenseRef-.+', re.UNICODE)
-        if (p[1] in config.LICENSE_MAP.keys()) or (ref_re.match(p[1]) is not None):
+        if (p[1].encode("utf-8") in config.LICENSE_MAP.keys()) or (ref_re.match(p[1]) is not None):
```
Do not encode here but at the boundaries

pombredanne commented on tests/test_document.py:
```diff
@@ -71,14 +72,14 @@ def test_license_list(self):
     def test_from_full_name(self):
         mit = License.from_full_name('MIT License')
-        assert mit.identifier == 'MIT'
+        assert mit.identifier == 'MIT'.decode("utf-8")
```
Do not decode or encode here but at the boundaries

pombredanne commented on tests/test_document.py~:
```diff
@@ -0,0 +1,85 @@
+# Copyright (c) 2014 Ahmed H. Ismail
```
do not commit editor backup files like this file

pombredanne commented on tests/test_document.py:
```diff
@@ -72,12 +72,12 @@ def test_license_list(self):
     def test_from_full_name(self):
         mit = License.from_full_name('MIT License')
-        assert mit.identifier == 'MIT'.decode("utf-8")
+        assert mit.identifier.decode("utf-8") == 'MIT'
```
Do not decode or encode here but at the boundaries

pombredanne commented on tests/test_document.py:
```diff
@@ -72,12 +72,12 @@ def test_license_list(self):
     def test_from_identifier(self):
         mit = License.from_identifier('MIT')
-        assert mit.full_name.encode("utf-8") == 'MIT License'.encode("utf-8")
+        assert mit.full_name == 'MIT License'.encode("utf-8")
```
Do not decode or encode here but at the boundaries

rtgdk commented:
@pombredanne I will fix all those things once I have a basic idea where these tests are failing. I know we should encode and decode at the boundaries but I just wanted a look where these implicit conversion are failing. About that commit message I know that commit messages should be clear and concise but I was just testing some changes to encoding and decoding. I plan to squash all the commits so these commit messages were just a way to push to github fast.

Apart from that, I have succesfully passed Python2 tests. But there is a new error in Python3 which never occured before in any of the test commit I did. And I don't know where I am going wrong.
Here is the code from `test_document.py` 

>  assert mit.identifier == 'MIT'

This is throwing an assertionError. i have tried decoding/encoding wherever possible but can't seem to figure out the error in Python3. **Please help me with this.**

pombredanne commented:
To solve your problem, you should never have to encode or decode. Just delete any of these. Just ensure instead that whenever you read or write something in and out you only return a proper str  (which will be always unicode since you are either Py3 or using Py2 with unicode_literals.) and never bytes

rtgdk commented:
@pombredanne  Yes that was my basic idea after reading about unicode changes in python2 and 3 on the web. https://nedbatchelder.com/text/unipain.html This article too said the same thing.
I did plan to remove encode/decode once I knew where the error was occuring (thats why all these commits ) and add them at the boundaries.
Maybe this was causing assertion error.
I'll rectify it as soon as possible.

pombredanne commented:
@nedbat 's article is an excellent reference. Ned is also super helpful and active on the #python IRC channel on Freenode with the nedbat nick.

pombredanne commented:
Though he is US east coast time.

rtgdk commented:
@pombredanne  Finally after more than 40 commits, I think I have fixed the existing test. Once you review the final changes, I'll squash the commits.
I have learned a lot about unicode, their conversions, their behaviour when reading/writing in file in the last few hours. Thanks to you.
I hope you like these changes.

rtgdk commented:
@pombredanne  Did you have a look at the changes ?
As you suggested, decoding/encoding at the boundaries. So I added unicode conversion when the data is being stored in the dictionary rather than opening the file with `codecs with utf8 encoding` as this was causing error with `csv` module. 
Apart from that I have added try except once while reading` '\n'.` Either this way or I can open the initial file with codecs and then remove all the decoding that being done in parsers.tagvalue.py.

pombredanne commented:
@rtgdk Consider using `unicodecsv` from Pypi then: the stdlib `csv` does not support unicode. This would make the solution much cleaner

pombredanne commented:
Using codecs is the way to go IMHO. Then fixing any issue that will come up

rtgdk commented:
On it. Squashing all the previous commits.

rtgdk commented:
@pombredanne  Tests cannot import `unicodecsv`. It's working fine in terminal. What to do ?
Also it's written in the official documentation of unicodecsv that there is no need to use `codecs` wrapper since unicodecsv expect bytestream and not unicode. So what to do? Use `open` only ?

rtgdk commented:
@pombredanne  Did you have a look at the tests? What should I do ?

rtgdk commented:
Travis and CI tests are failing. They cannot import unicodecsv. When I tried it out locally, it seemed to work without any additionally library download.

pombredanne commented:
What happens if you clean your venv, create a clean one and then run `python setup.py install` in the vend and then run the tests with  `py.test -vvs -x`?

pombredanne commented on spdx/config.py~:
```diff
@@ -0,0 +1,44 @@
+
```
Do NOT commit ~ files

pombredanne commented on spdx/config.py:
```diff
@@ -12,22 +12,22 @@
+from __future__ import unicode_literals
 
-import csv
+import unicodecsv as csv
```
Do not use `as` here.

pombredanne commented on spdx/config.py:
```diff
@@ -23,15 +23,11 @@
-		except NameError :
-			dict.__setitem__(self,key,value)
-			dict.__setitem__(self,value,key)
+		dict.__setitem__(self, key,value)
```
Is the spacing 4 spaces?

pombredanne commented on README.md:
```diff
@@ -1,5 +1,15 @@
-Python SPDX Library to parse, validate and create SPDX documents
-==========================
+# Python SPDX Library to parse, validate and create SPDX documents
```
How is this related to this PR?

pombredanne replied:
?

pombredanne commented on spdx/config.py:
```diff
@@ -12,22 +12,22 @@
+	def __setitem__(self, key, value):
+		dict.__setitem__(self, key,value)
+		dict.__setitem__(self, value,key)
+	def __delitem__(self, key):
```
Why removing the empty line? And what about 4 spaces indent?

rtgdk replied:
I may have missed it. I use tab for indentation which is set to default 4 spaces. Should I use 2 spaces ? I think the whole code is written with 4 spaces indentation.

pombredanne commented on spdx/parsers/lexers/tagvalue.py:
```diff
@@ -95,7 +96,11 @@ def t_text_end(self, t):
             t.lexer.text_start:t.lexer.lexpos].strip()
-        t.lexer.lineno += t.value.count('\n')
+        try :
+        	unicode
```
What is unicode meaning here? This does not make sense.

rtgdk replied:
It will raise an error in Python3. So the exception will be executed.

pombredanne commented on spdx/parsers/lexers/tagvalue.py:
```diff
@@ -95,7 +96,11 @@ def t_text_end(self, t):
-        t.lexer.lineno += t.value.count('\n')
+        try :
+        	unicode
+        	t.lexer.lineno += t.value.count('\n'.encode("utf-8"))
```
Why are you encoding to bytes here?

rtgdk replied:
Actually tagvalue.py in parsers has a lot of decode options, which was raising error while opening the file with `codecs`. So I plan to not change the current test ( openfile with `open` ) and change the code only once here.  
Default encoding is ascii in `open`, so I have to specifically encode it to utf-8. This is needed to pass the test I think.

pombredanne replied:
If this is needed to pass the test, then may be the tests need to be updated instead?

pombredanne commented:
Hint: add unicodecsv to setup.py's `install_requires`

pombredanne commented on setup.py~:
```diff
@@ -0,0 +1,39 @@
+#!/usr/bin/env python
```
Please, do not commit editor backup files...

rtgdk replied:
I am really sorry. I plan to remove all those backup files and write correct commit message once I get the correct solution. These are just for fast git push to github. I'll squash all the commits once I'm done.

pombredanne commented on spdx/config.py:
```diff
@@ -32,7 +33,7 @@ def __delitem__(self, key):
 def load_license_list():
     FILE_NAME = os.path.join(os.path.dirname(__file__), 'spdx_licenselist.csv')
-    with open(FILE_NAME, 'r') as file_in:
+    with codecs.open(FILE_NAME, 'r', encoding="utf-8") as file_in:
```
I know it is not your doing, but improving the code while you touch it is a good idea.  FILE_NAME and uppercase names are used for global constants by convention and not scoped variables. Use `file_name` instead.

rtgdk replied:
Ok I'll do that.

pombredanne commented on spdx/config.py~:
```diff
@@ -32,8 +33,8 @@ class TwoWayDict(dict):
     FILE_NAME = os.path.join(os.path.dirname(__file__), 'spdx_licenselist.csv')
-    with open(FILE_NAME, 'r') as file_in:
-        reader = csv.DictReader(file_in)
+    with codecs.open(FILE_NAME, 'r', encoding="utf-8") as file_in:
```
FILE_NAME and uppercase names are used for global constants by convention and not scoped variables. Use `file_name` instead.

rtgdk commented:
@pombredanne I have a doubt. While using unicodecsv, the module is gonna convert bytes to unicode ( See. So it expects data to be in bytes. But in python3, data would be in str. So while decoding, it will raise error. ( See travis test : https://travis-ci.org/spdx/tools-python/jobs/215542055 ).
What should I do now. Convert file to byte stream ?

pombredanne commented:
@rtgdk If you have some issues with unicodecsv I suggest filing a bug with them. The best way would be to write a test that shows the issue first.

pombredanne commented:
@rtgdk note also that you will need to rebase your PR on the latest master

pombredanne commented:
@rtgdk It sounds like you merged and did not rebase? or something else?

pombredanne commented on README.md:
```diff
@@ -1,5 +1,15 @@
-==========================
+# Python SPDX Library to parse, validate and create SPDX documents
+
+| Linux | macOS | Windows |
```
How come these files are part of your PR? I am a tad confused.

pombredanne replied:
?

rtgdk commented:
@pombredanne  Sorry, for not getting back to this earlier. My exams were going on.
I have currently written a simple test to show `unicodecsv` error in Python3.
I will rebase and merge with the master branch as soon as I make the changes in `config.py` to read the file using `csv`.

rtgdk commented:
@pombredanne  I have made the `unicodecsv` test. It is working in python2 but not in Python3.( That is why Travis test failed)

Also instead of `unicodecsv` I am using `csv` module for compatibility with both 2 and 3 version.

pombredanne commented:
@rtgdk Thanks. Let me check that after the week end!

rtgdk commented:
@pombredanne  Did you have a look at the changes ?

pombredanne commented:
@rtgdk I had looked at you changes... but did you check all the comments I made while reviewing the code?

pombredanne commented on spdx/config.py:
```diff
@@ -21,18 +21,18 @@
-    def __delitem__(self, key):
-        dict.__delitem__(self, self[key])
-        dict.__delitem__(self, key)
+	def __setitem__(self, key, value):
```
why did you change the indenting here?

pombredanne commented on tests/test_unicodecsv.py:
```diff
@@ -0,0 +1,21 @@
+    def test_unicode(self):
+    	dct=dict()
+    	with open('data/sample.csv', 'r') as file_in:
+		    reader = unicodecsv.DictReader(file_in)
```
Please use 4 spaces for indent, not 8

pombredanne commented on tests/test_unicodecsv.py:
```diff
@@ -0,0 +1,21 @@
+
+class TestUnicodeCsv(unittest.TestCase):
+
+    def test_unicode(self):
```
I am not sure we care for a test of unicodecsv. What we need is a test of the support of unicode in thsi spdx library instead. For inatnce, what if we pass unicode content? etc.

pombredanne commented:
@rtgdk Thanks. You still have some tests that are failing though.

rtgdk commented:
@pombredanne  Sorry for not getting back to this earlier. My exams were going on.
The tests are failing for Python3 as the key-value pair is in unicode ( `unicode_literals `). But passing the key as byte in Python3 will create `Key Error `as no such key is present in the dictionary.
I'll just change it to `assert not` in case of Python3.

pombredanne commented:
Hum ok. but I cannot fathom why there would be a different behavior there.... Also did you check the other comments I made on the PR? No rush if you have exams of course :)

rtgdk commented:
![screenshot from 2017-05-02 20-22-18](https://cloud.githubusercontent.com/assets/17159160/25624108/1032dd38-2f76-11e7-8007-d5dab17381ec.png)
This is why Python3 shows error. Since we are using unicode_literals, all the data is stored in unicode. So accessing them in bytes in not possible in Python3.

![screenshot from 2017-05-02 20-23-13](https://cloud.githubusercontent.com/assets/17159160/25624180/3df87174-2f76-11e7-93db-0950e44a9eea.png) 
And this is case in Python2

rtgdk commented:
I have now left two empty lines after imports.
README.md is still need to be change. But I can't seem to figure out why this file is being shown in the changes. I have taken the exact same file from spdx:master. You can see in the above screenshot that there is a field for Linux,macOS and Windows. But there is no such field in the original file. So I don't know what to do.

pombredanne commented:
@rtgdk I will review and merge, and eventually cherry picking the commits as needed. Thanks!

rtgdk commented:
@pombredanne  What about this PR ? Any changes needed ?

pombredanne commented:
I just need to review it .... sorry for taking some time

pombredanne commented:
Actually this has all been merged in alright after a rebase some while ago.... with https://github.com/spdx/tools-python/commit/dce58f5d30c792c3d582b431accfb402f175dd83  so closing
